sound/soc/sof/sof-audio.c

```diff
@@ -11,6 +11,66 @@
 #include "sof-audio.h"
 #include "ops.h"
 
+static int sof_kcontrol_setup(struct snd_sof_dev *sdev, struct snd_sof_control *scontrol)
+{
+	int ipc_cmd, ctrl_type;
+	int ret;
+
+	/* reset readback offset for scontrol */
+	scontrol->readback_offset = 0;
+
+	/* notify DSP of kcontrol values */
+	switch (scontrol->cmd) {
+	case SOF_CTRL_CMD_VOLUME:
+	case SOF_CTRL_CMD_ENUM:
+	case SOF_CTRL_CMD_SWITCH:
+		ipc_cmd = SOF_IPC_COMP_SET_VALUE;
+		ctrl_type = SOF_CTRL_TYPE_VALUE_CHAN_SET;
+		break;
+	case SOF_CTRL_CMD_BINARY:
+		ipc_cmd = SOF_IPC_COMP_SET_DATA;
+		ctrl_type = SOF_CTRL_TYPE_DATA_SET;
+		break;
+	default:
+		return 0;
+	}
+
+	ret = snd_sof_ipc_set_get_comp_data(scontrol, ipc_cmd, ctrl_type, scontrol->cmd, true);
+	if (ret < 0)
+		dev_err(sdev->dev, "error: failed kcontrol value set for widget: %d\n",
+			scontrol->comp_id);
+
+	return ret;
+}
+
+static int sof_dai_config_setup(struct snd_sof_dev *sdev, struct snd_sof_dai *dai)
+{
+	struct sof_ipc_reply reply;
+	struct sof_ipc_dai_config *config = dai->dai_config;
+	int ret;
+
+	if (!config) {
+		dev_err(sdev->dev, "error: no config for DAI %s\n", dai->name);
+		return -EINVAL;
+	}
+
+	/*
+	 * The link DMA channel would be invalidated for running streams but not for streams that
+	 * were in the PAUSED state during suspend. So invalidate it here before setting the dai
+	 * config in the DSP.
+	 */
+	if (config->type == SOF_DAI_INTEL_HDA)
+		config->hda.link_dma_ch = DMA_CHAN_INVALID;
```

Comment:
it's quite ugly to have this sort of Intel-specific code in the core. can we use some sort of abstraction here?

I understand it was there before your patch but that's concerning that we let this go though.

Reply:
this piece of code is eventually removed by the last patch in the PR. This patch just moves code around to facilitate that eventually.